Added Support for TP-link C1200

swwgames opened the pull request:
Added Support for TP-link C1200
The main difference is the login process, the C1200 doesn't expose its encryption keys.

AlexandrErohin commented on README.md:
````diff
@@ -68,6 +68,62 @@ try:
             router.logout()  # always logout as TP-Link Web Interface only supports upto 1 user logged
 ```
+If you have the TP-link C1200 V2, you can use the TplinkC1200Router class instead of the TplinkRouter class.
+Remember that the password is different, here you need to use the web encrypted password.
````
Is there any chance to get password encryption algorithm from JavaScript? Would be great if user can only pass the real password and the client can encrypt it

swwgames replied:
I am sorry, I tried to get the password encryption but this is way to hard for me.
i found that calling http://192.168.0.1/cgi-bin/luci/;stok=/login?form=cloud_login&operation=read
returns: 
```
{"success":true,"data":{"username":"","password":["D1E79FF135D14E342D76185C23024E6DEAD4D6EC2C317A526C811E83538EA4E5ED8E1B0EEE5CE26E3C1B6A5F1FE11FA804F28B7E8821CA90AFA5B2F300DF99FDA27C9D2131E031EA11463C47944C05005EF4C1CE932D7F4A87C7563581D9F27F0C305023FCE94997EC7D790696E784357ED803A610EBB71B12A8BE5936429BFD","010001"]}}
```
this is needed in the _request_pwd function.
But i can't find a api call for the seq key, this is needed in the _request_seq function.
I suspect that this is generated in the javascript code.
This is the encrypt.js file which is called to encrypt the password. val is the login password and param is the list i put above.
```js
(function($) {
    $.su = $.su || {};
    $.su.encrypt = function(val, param) {
        var dbits;
        var canary = 244837814094590;
        var j_lm = ((canary & 16777215) == 15715070);
        function BigInteger(a, b, c) {
            if (a != null) {
                if ("number" == typeof a) {
                    this.fromNumber(a, b, c)
                } else {
                    if (b == null && "string" != typeof a) {
                        this.fromString(a, 256)
                    } else {
                        this.fromString(a, b)
                    }
                }
            }
        }
        function nbi() {
            return new BigInteger(null)
        }
        function am1(i, x, w, j, c, n) {
            while (--n >= 0) {
                var v = x * this[i++] + w[j] + c;
                c = Math.floor(v / 67108864);
                w[j++] = v & 67108863
            }
            return c
        }
        function am2(i, x, w, j, c, n) {
            var xl = x & 32767
              , xh = x >> 15;
            while (--n >= 0) {
                var l = this[i] & 32767;
                var h = this[i++] >> 15;
                var m = xh * l + h * xl;
                l = xl * l + ((m & 32767) << 15) + w[j] + (c & 1073741823);
                c = (l >>> 30) + (m >>> 15) + xh * h + (c >>> 30);
                w[j++] = l & 1073741823
            }
            return c
        }
        function am3(i, x, w, j, c, n) {
            var xl = x & 16383
              , xh = x >> 14;
            while (--n >= 0) {
                var l = this[i] & 16383;
                var h = this[i++] >> 14;
                var m = xh * l + h * xl;
                l = xl * l + ((m & 16383) << 14) + w[j] + c;
                c = (l >> 28) + (m >> 14) + xh * h;
                w[j++] = l & 268435455
            }
            return c
        }
        if (j_lm && (navigator.appName == "Microsoft Internet Explorer")) {
            BigInteger.prototype.am = am2;
            dbits = 30
        } else {
            if (j_lm && (navigator.appName != "Netscape")) {
                BigInteger.prototype.am = am1;
                dbits = 26
            } else {
                BigInteger.prototype.am = am3;
                dbits = 28
            }
        }
        BigInteger.prototype.DB = dbits;
        BigInteger.prototype.DM = ((1 << dbits) - 1);
        BigInteger.prototype.DV = (1 << dbits);
        var BI_FP = 52;
        BigInteger.prototype.FV = Math.pow(2, BI_FP);
        BigInteger.prototype.F1 = BI_FP - dbits;
        BigInteger.prototype.F2 = 2 * dbits - BI_FP;
        var BI_RM = "0123456789abcdefghijklmnopqrstuvwxyz";
        var BI_RC = new Array();
        var rr, vv;
        rr = "0".charCodeAt(0);
        for (vv = 0; vv <= 9; ++vv) {
            BI_RC[rr++] = vv
        }
        rr = "a".charCodeAt(0);
        for (vv = 10; vv < 36; ++vv) {
            BI_RC[rr++] = vv
        }
        rr = "A".charCodeAt(0);
        for (vv = 10; vv < 36; ++vv) {
            BI_RC[rr++] = vv
        }
        function int2char(n) {
            return BI_RM.charAt(n)
        }
        function intAt(s, i) {
            var c = BI_RC[s.charCodeAt(i)];
            return (c == null) ? -1 : c
        }
        function bnpCopyTo(r) {
            for (var i = this.t - 1; i >= 0; --i) {
                r[i] = this[i]
            }
            r.t = this.t;
            r.s = this.s
        }
        function bnpFromInt(x) {
            this.t = 1;
            this.s = (x < 0) ? -1 : 0;
            if (x > 0) {
                this[0] = x
            } else {
                if (x < -1) {
                    this[0] = x + this.DV
                } else {
                    this.t = 0
                }
            }
        }
        function nbv(i) {
            var r = nbi();
            r.fromInt(i);
            return r
        }
        function bnpFromString(s, b) {
            var k;
            if (b == 16) {
                k = 4
            } else {
                if (b == 8) {
                    k = 3
                } else {
                    if (b == 256) {
                        k = 8
                    } else {
                        if (b == 2) {
                            k = 1
                        } else {
                            if (b == 32) {
                                k = 5
                            } else {
                                if (b == 4) {
                                    k = 2
                                } else {
                                    this.fromRadix(s, b);
                                    return
                                }
                            }
                        }
                    }
                }
            }
            this.t = 0;
            this.s = 0;
            var i = s.length
              , mi = false
              , sh = 0;
            while (--i >= 0) {
                var x = (k == 8) ? s[i] & 255 : intAt(s, i);
                if (x < 0) {
                    if (s.charAt(i) == "-") {
                        mi = true
                    }
                    continue
                }
                mi = false;
                if (sh == 0) {
                    this[this.t++] = x
                } else {
                    if (sh + k > this.DB) {
                        this[this.t - 1] |= (x & ((1 << (this.DB - sh)) - 1)) << sh;
                        this[this.t++] = (x >> (this.DB - sh))
                    } else {
                        this[this.t - 1] |= x << sh
                    }
                }
                sh += k;
                if (sh >= this.DB) {
                    sh -= this.DB
                }
            }
            if (k == 8 && (s[0] & 128) != 0) {
                this.s = -1;
                if (sh > 0) {
                    this[this.t - 1] |= ((1 << (this.DB - sh)) - 1) << sh
                }
            }
            this.clamp();
            if (mi) {
                BigInteger.ZERO.subTo(this, this)
            }
        }
        function bnpClamp() {
            var c = this.s & this.DM;
            while (this.t > 0 && this[this.t - 1] == c) {
                --this.t
            }
        }
        function bnToString(b) {
            if (this.s < 0) {
                return "-" + this.negate().toString(b)
            }
            var k;
            if (b == 16) {
                k = 4
            } else {
                if (b == 8) {
                    k = 3
                } else {
                    if (b == 2) {
                        k = 1
                    } else {
                        if (b == 32) {
                            k = 5
                        } else {
                            if (b == 4) {
                                k = 2
                            } else {
                                return this.toRadix(b)
                            }
                        }
                    }
                }
            }
            var km = (1 << k) - 1, d, m = false, r = "", i = this.t;
            var p = this.DB - (i * this.DB) % k;
            if (i-- > 0) {
                if (p < this.DB && (d = this[i] >> p) > 0) {
                    m = true;
                    r = int2char(d)
                }
                while (i >= 0) {
                    if (p < k) {
                        d = (this[i] & ((1 << p) - 1)) << (k - p);
                        d |= this[--i] >> (p += this.DB - k)
                    } else {
                        d = (this[i] >> (p -= k)) & km;
                        if (p <= 0) {
                            p += this.DB;
                            --i
                        }
                    }
                    if (d > 0) {
                        m = true
                    }
                    if (m) {
                        r += int2char(d)
                    }
                }
            }
            return m ? r : "0"
        }
        function bnNegate() {
            var r = nbi();
            BigInteger.ZERO.subTo(this, r);
            return r
        }
        function bnAbs() {
            return (this.s < 0) ? this.negate() : this
        }
        function bnCompareTo(a) {
            var r = this.s - a.s;
            if (r != 0) {
                return r
            }
            var i = this.t;
            r = i - a.t;
            if (r != 0) {
                return (this.s < 0) ? -r : r
            }
            while (--i >= 0) {
                if ((r = this[i] - a[i]) != 0) {
                    return r
                }
            }
            return 0
        }
        function nbits(x) {
            var r = 1, t;
            if ((t = x >>> 16) != 0) {
                x = t;
                r += 16
            }
            if ((t = x >> 8) != 0) {
                x = t;
                r += 8
            }
            if ((t = x >> 4) != 0) {
                x = t;
                r += 4
            }
            if ((t = x >> 2) != 0) {
                x = t;
                r += 2
            }
            if ((t = x >> 1) != 0) {
                x = t;
                r += 1
            }
            return r
        }
        function bnBitLength() {
            if (this.t <= 0) {
                return 0
            }
            return this.DB * (this.t - 1) + nbits(this[this.t - 1] ^ (this.s & this.DM))
        }
        function bnpDLShiftTo(n, r) {
            var i;
            for (i = this.t - 1; i >= 0; --i) {
                r[i + n] = this[i]
            }
            for (i = n - 1; i >= 0; --i) {
                r[i] = 0
            }
            r.t = this.t + n;
            r.s = this.s
        }
        function bnpDRShiftTo(n, r) {
            for (var i = n; i < this.t; ++i) {
                r[i - n] = this[i]
            }
            r.t = Math.max(this.t - n, 0);
            r.s = this.s
        }
        function bnpLShiftTo(n, r) {
            var bs = n % this.DB;
            var cbs = this.DB - bs;
            var bm = (1 << cbs) - 1;
            var ds = Math.floor(n / this.DB), c = (this.s << bs) & this.DM, i;
            for (i = this.t - 1; i >= 0; --i) {
                r[i + ds + 1] = (this[i] >> cbs) | c;
                c = (this[i] & bm) << bs
            }
            for (i = ds - 1; i >= 0; --i) {
                r[i] = 0
            }
            r[ds] = c;
            r.t = this.t + ds + 1;
            r.s = this.s;
            r.clamp()
        }
        function bnpRShiftTo(n, r) {
            r.s = this.s;
            var ds = Math.floor(n / this.DB);
            if (ds >= this.t) {
                r.t = 0;
                return
            }
            var bs = n % this.DB;
            var cbs = this.DB - bs;
            var bm = (1 << bs) - 1;
            r[0] = this[ds] >> bs;
            for (var i = ds + 1; i < this.t; ++i) {
                r[i - ds - 1] |= (this[i] & bm) << cbs;
                r[i - ds] = this[i] >> bs
            }
            if (bs > 0) {
                r[this.t - ds - 1] |= (this.s & bm) << cbs
            }
            r.t = this.t - ds;
            r.clamp()
        }
        function bnpSubTo(a, r) {
            var i = 0
              , c = 0
              , m = Math.min(a.t, this.t);
            while (i < m) {
                c += this[i] - a[i];
                r[i++] = c & this.DM;
                c >>= this.DB
            }
            if (a.t < this.t) {
                c -= a.s;
                while (i < this.t) {
                    c += this[i];
                    r[i++] = c & this.DM;
                    c >>= this.DB
                }
                c += this.s
            } else {
                c += this.s;
                while (i < a.t) {
                    c -= a[i];
                    r[i++] = c & this.DM;
                    c >>= this.DB
                }
                c -= a.s
            }
            r.s = (c < 0) ? -1 : 0;
            if (c < -1) {
                r[i++] = this.DV + c
            } else {
                if (c > 0) {
                    r[i++] = c
                }
            }
            r.t = i;
            r.clamp()
        }
        function bnpMultiplyTo(a, r) {
            var x = this.abs()
              , y = a.abs();
            var i = x.t;
            r.t = i + y.t;
            while (--i >= 0) {
                r[i] = 0
            }
            for (i = 0; i < y.t; ++i) {
                r[i + x.t] = x.am(0, y[i], r, i, 0, x.t)
            }
            r.s = 0;
            r.clamp();
            if (this.s != a.s) {
                BigInteger.ZERO.subTo(r, r)
            }
        }
        function bnpSquareTo(r) {
            var x = this.abs();
            var i = r.t = 2 * x.t;
            while (--i >= 0) {
                r[i] = 0
            }
            for (i = 0; i < x.t - 1; ++i) {
                var c = x.am(i, x[i], r, 2 * i, 0, 1);
                if ((r[i + x.t] += x.am(i + 1, 2 * x[i], r, 2 * i + 1, c, x.t - i - 1)) >= x.DV) {
                    r[i + x.t] -= x.DV;
                    r[i + x.t + 1] = 1
                }
            }
            if (r.t > 0) {
                r[r.t - 1] += x.am(i, x[i], r, 2 * i, 0, 1)
            }
            r.s = 0;
            r.clamp()
        }
        function bnpDivRemTo(m, q, r) {
            var pm = m.abs();
            if (pm.t <= 0) {
                return
            }
            var pt = this.abs();
            if (pt.t < pm.t) {
                if (q != null) {
                    q.fromInt(0)
                }
                if (r != null) {
                    this.copyTo(r)
                }
                return
            }
            if (r == null) {
                r = nbi()
            }
            var y = nbi()
              , ts = this.s
              , ms = m.s;
            var nsh = this.DB - nbits(pm[pm.t - 1]);
            if (nsh > 0) {
                pm.lShiftTo(nsh, y);
                pt.lShiftTo(nsh, r)
            } else {
                pm.copyTo(y);
                pt.copyTo(r)
            }
            var ys = y.t;
            var y0 = y[ys - 1];
            if (y0 == 0) {
                return
            }
            var yt = y0 * (1 << this.F1) + ((ys > 1) ? y[ys - 2] >> this.F2 : 0);
            var d1 = this.FV / yt
              , d2 = (1 << this.F1) / yt
              , e = 1 << this.F2;
            var i = r.t
              , j = i - ys
              , t = (q == null) ? nbi() : q;
            y.dlShiftTo(j, t);
            if (r.compareTo(t) >= 0) {
                r[r.t++] = 1;
                r.subTo(t, r)
            }
            BigInteger.ONE.dlShiftTo(ys, t);
            t.subTo(y, y);
            while (y.t < ys) {
                y[y.t++] = 0
            }
            while (--j >= 0) {
                var qd = (r[--i] == y0) ? this.DM : Math.floor(r[i] * d1 + (r[i - 1] + e) * d2);
                if ((r[i] += y.am(0, qd, r, j, 0, ys)) < qd) {
                    y.dlShiftTo(j, t);
                    r.subTo(t, r);
                    while (r[i] < --qd) {
                        r.subTo(t, r)
                    }
                }
            }
            if (q != null) {
                r.drShiftTo(ys, q);
                if (ts != ms) {
                    BigInteger.ZERO.subTo(q, q)
                }
            }
            r.t = ys;
            r.clamp();
            if (nsh > 0) {
                r.rShiftTo(nsh, r)
            }
            if (ts < 0) {
                BigInteger.ZERO.subTo(r, r)
            }
        }
        function bnMod(a) {
            var r = nbi();
            this.abs().divRemTo(a, null, r);
            if (this.s < 0 && r.compareTo(BigInteger.ZERO) > 0) {
                a.subTo(r, r)
            }
            return r
        }
        function Classic(m) {
            this.m = m
        }
        function cConvert(x) {
            if (x.s < 0 || x.compareTo(this.m) >= 0) {
                return x.mod(this.m)
            } else {
                return x
            }
        }
        function cRevert(x) {
            return x
        }
        function cReduce(x) {
            x.divRemTo(this.m, null, x)
        }
        function cMulTo(x, y, r) {
            x.multiplyTo(y, r);
            this.reduce(r)
        }
        function cSqrTo(x, r) {
            x.squareTo(r);
            this.reduce(r)
        }
        Classic.prototype.convert = cConvert;
        Classic.prototype.revert = cRevert;
        Classic.prototype.reduce = cReduce;
        Classic.prototype.mulTo = cMulTo;
        Classic.prototype.sqrTo = cSqrTo;
        function bnpInvDigit() {
            if (this.t < 1) {
                return 0
            }
            var x = this[0];
            if ((x & 1) == 0) {
                return 0
            }
            var y = x & 3;
            y = (y * (2 - (x & 15) * y)) & 15;
            y = (y * (2 - (x & 255) * y)) & 255;
            y = (y * (2 - (((x & 65535) * y) & 65535))) & 65535;
            y = (y * (2 - x * y % this.DV)) % this.DV;
            return (y > 0) ? this.DV - y : -y
        }
        function Montgomery(m) {
            this.m = m;
            this.mp = m.invDigit();
            this.mpl = this.mp & 32767;
            this.mph = this.mp >> 15;
            this.um = (1 << (m.DB - 15)) - 1;
            this.mt2 = 2 * m.t
        }
        function montConvert(x) {
            var r = nbi();
            x.abs().dlShiftTo(this.m.t, r);
            r.divRemTo(this.m, null, r);
            if (x.s < 0 && r.compareTo(BigInteger.ZERO) > 0) {
                this.m.subTo(r, r)
            }
            return r
        }
        function montRevert(x) {
            var r = nbi();
            x.copyTo(r);
            this.reduce(r);
            return r
        }
        function montReduce(x) {
            while (x.t <= this.mt2) {
                x[x.t++] = 0
            }
            for (var i = 0; i < this.m.t; ++i) {
                var j = x[i] & 32767;
                var u0 = (j * this.mpl + (((j * this.mph + (x[i] >> 15) * this.mpl) & this.um) << 15)) & x.DM;
                j = i + this.m.t;
                x[j] += this.m.am(0, u0, x, i, 0, this.m.t);
                while (x[j] >= x.DV) {
                    x[j] -= x.DV;
                    x[++j]++
                }
            }
            x.clamp();
            x.drShiftTo(this.m.t, x);
            if (x.compareTo(this.m) >= 0) {
                x.subTo(this.m, x)
            }
        }
        function montSqrTo(x, r) {
            x.squareTo(r);
            this.reduce(r)
        }
        function montMulTo(x, y, r) {
            x.multiplyTo(y, r);
            this.reduce(r)
        }
        Montgomery.prototype.convert = montConvert;
        Montgomery.prototype.revert = montRevert;
        Montgomery.prototype.reduce = montReduce;
        Montgomery.prototype.mulTo = montMulTo;
        Montgomery.prototype.sqrTo = montSqrTo;
        function bnpIsEven() {
            return ((this.t > 0) ? (this[0] & 1) : this.s) == 0
        }
        function bnpExp(e, z) {
            if (e > 4294967295 || e < 1) {
                return BigInteger.ONE
            }
            var r = nbi()
              , r2 = nbi()
              , g = z.convert(this)
              , i = nbits(e) - 1;
            g.copyTo(r);
            while (--i >= 0) {
                z.sqrTo(r, r2);
                if ((e & (1 << i)) > 0) {
                    z.mulTo(r2, g, r)
                } else {
                    var t = r;
                    r = r2;
                    r2 = t
                }
            }
            return z.revert(r)
        }
        function bnModPowInt(e, m) {
            var z;
            if (e < 256 || m.isEven()) {
                z = new Classic(m)
            } else {
                z = new Montgomery(m)
            }
            return this.exp(e, z)
        }
        BigInteger.prototype.copyTo = bnpCopyTo;
        BigInteger.prototype.fromInt = bnpFromInt;
        BigInteger.prototype.fromString = bnpFromString;
        BigInteger.prototype.clamp = bnpClamp;
        BigInteger.prototype.dlShiftTo = bnpDLShiftTo;
        BigInteger.prototype.drShiftTo = bnpDRShiftTo;
        BigInteger.prototype.lShiftTo = bnpLShiftTo;
        BigInteger.prototype.rShiftTo = bnpRShiftTo;
        BigInteger.prototype.subTo = bnpSubTo;
        BigInteger.prototype.multiplyTo = bnpMultiplyTo;
        BigInteger.prototype.squareTo = bnpSquareTo;
        BigInteger.prototype.divRemTo = bnpDivRemTo;
        BigInteger.prototype.invDigit = bnpInvDigit;
        BigInteger.prototype.isEven = bnpIsEven;
        BigInteger.prototype.exp = bnpExp;
        BigInteger.prototype.toString = bnToString;
        BigInteger.prototype.negate = bnNegate;
        BigInteger.prototype.abs = bnAbs;
        BigInteger.prototype.compareTo = bnCompareTo;
        BigInteger.prototype.bitLength = bnBitLength;
        BigInteger.prototype.mod = bnMod;
        BigInteger.prototype.modPowInt = bnModPowInt;
        BigInteger.ZERO = nbv(0);
        BigInteger.ONE = nbv(1);
        function Arcfour() {
            this.i = 0;
            this.j = 0;
            this.S = new Array()
        }
        function ARC4init(key) {
            var i, j, t;
            for (i = 0; i < 256; ++i) {
                this.S[i] = i
            }
            j = 0;
            for (i = 0; i < 256; ++i) {
                j = (j + this.S[i] + key[i % key.length]) & 255;
                t = this.S[i];
                this.S[i] = this.S[j];
                this.S[j] = t
            }
            this.i = 0;
            this.j = 0
        }
        function ARC4next() {
            var t;
            this.i = (this.i + 1) & 255;
            this.j = (this.j + this.S[this.i]) & 255;
            t = this.S[this.i];
            this.S[this.i] = this.S[this.j];
            this.S[this.j] = t;
            return this.S[(t + this.S[this.i]) & 255]
        }
        Arcfour.prototype.init = ARC4init;
        Arcfour.prototype.next = ARC4next;
        function prng_newstate() {
            return new Arcfour()
        }
        var rng_psize = 256;
        var rng_state;
        var rng_pool;
        var rng_pptr;
        function rng_seed_int(x) {
            rng_pool[rng_pptr++] ^= x & 255;
            rng_pool[rng_pptr++] ^= (x >> 8) & 255;
            rng_pool[rng_pptr++] ^= (x >> 16) & 255;
            rng_pool[rng_pptr++] ^= (x >> 24) & 255;
            if (rng_pptr >= rng_psize) {
                rng_pptr -= rng_psize
            }
        }
        function rng_seed_time() {
            rng_seed_int(new Date().getTime())
        }
        if (rng_pool == null) {
            rng_pool = new Array();
            rng_pptr = 0;
            var t;
            if (window.crypto && window.crypto.getRandomValues) {
                var ua = new Uint8Array(32);
                window.crypto.getRandomValues(ua);
                for (t = 0; t < 32; ++t) {
                    rng_pool[rng_pptr++] = ua[t]
                }
            }
            if (navigator.appName == "Netscape" && navigator.appVersion < "5" && window.crypto) {
                var z = window.crypto.random(32);
                for (t = 0; t < z.length; ++t) {
                    rng_pool[rng_pptr++] = z.charCodeAt(t) & 255
                }
            }
            while (rng_pptr < rng_psize) {
                t = Math.floor(65536 * Math.random());
                rng_pool[rng_pptr++] = t >>> 8;
                rng_pool[rng_pptr++] = t & 255
            }
            rng_pptr = 0;
            rng_seed_time()
        }
        function rng_get_byte() {
            if (rng_state == null) {
                rng_seed_time();
                rng_state = prng_newstate();
                rng_state.init(rng_pool);
                for (rng_pptr = 0; rng_pptr < rng_pool.length; ++rng_pptr) {
                    rng_pool[rng_pptr] = 0
                }
                rng_pptr = 0
            }
            return rng_state.next()
        }
        function rng_get_bytes(ba) {
            var i;
            for (i = 0; i < ba.length; ++i) {
                ba[i] = rng_get_byte()
            }
        }
        function SecureRandom() {}
        SecureRandom.prototype.nextBytes = rng_get_bytes;
        function parseBigInt(str, r) {
            return new BigInteger(str,r)
        }
        function byte2Hex(b) {
            if (b < 16) {
                return "0" + b.toString(16)
            } else {
                return b.toString(16)
            }
        }
        function RSAKey() {
            this.n = null;
            this.e = 0;
            this.d = null;
            this.p = null;
            this.q = null;
            this.dmp1 = null;
            this.dmq1 = null;
            this.coeff = null
        }
        function RSASetPublic(N, E) {
            if (N != null && E != null && N.length > 0 && E.length > 0) {
                this.n = parseBigInt(N, 16);
                this.e = parseInt(E, 16)
            } else {
                alert("Invalid RSA public key")
            }
        }
        function RSADoPublic(x) {
            return x.modPowInt(this.e, this.n)
        }
        function nopadding(s, n) {
            if (n < s.length) {
                alert("Message too long for RSA");
                return null
            }
            var ba = new Array();
            var i = 0;
            var j = 0;
            while (i < s.length && j < n) {
                var c = s.charCodeAt(i++);
                if (c < 128) {
                    ba[j++] = c
                } else {
                    if ((c > 127) && (c < 2048)) {
                        ba[j++] = (c & 63) | 128;
                        ba[j++] = (c >> 6) | 192
                    } else {
                        ba[j++] = (c & 63) | 128;
                        ba[j++] = ((c >> 6) & 63) | 128;
                        ba[j++] = (c >> 12) | 224
                    }
                }
            }
            while (j < n) {
                ba[j++] = 0
            }
            return new BigInteger(ba)
        }
        function RSAEncrypt(text) {
            var m = nopadding(text, (this.n.bitLength() + 7) >> 3);
            if (m == null) {
                return null
            }
            var c = this.doPublic(m);
            if (c == null) {
                return null
            }
            var h = c.toString(16);
            if ((h.length & 1) == 0) {
                return h
            } else {
                return "0" + h
            }
        }
        RSAKey.prototype.doPublic = RSADoPublic;
        RSAKey.prototype.setPublic = RSASetPublic;
        RSAKey.prototype.encrypt = RSAEncrypt;
        var rsaObj = new RSAKey();
        var n = param[0];
        var e = param[1];
        rsaObj.setPublic(n, e);
        var result = rsaObj.encrypt(val);
        if (result.length != 256) {
            var l = Math.abs(256 - result.length);
            for (var i = 0; i < l; i++) {
                result = "0" + result
            }
        }
        return result
    }
    ;
    $.su.des = function(key, message, encrypt, mode, iv, padding) {
        if (encrypt) {
            message = unescape(encodeURIComponent(message))
        }
        var spfunction1 = new Array(16843776,0,65536,16843780,16842756,66564,4,65536,1024,16843776,16843780,1024,16778244,16842756,16777216,4,1028,16778240,16778240,66560,66560,16842752,16842752,16778244,65540,16777220,16777220,65540,0,1028,66564,16777216,65536,16843780,4,16842752,16843776,16777216,16777216,1024,16842756,65536,66560,16777220,1024,4,16778244,66564,16843780,65540,16842752,16778244,16777220,1028,66564,16843776,1028,16778240,16778240,0,65540,66560,0,16842756);
        var spfunction2 = new Array(-2146402272,-2147450880,32768,1081376,1048576,32,-2146435040,-2147450848,-2147483616,-2146402272,-2146402304,-2147483648,-2147450880,1048576,32,-2146435040,1081344,1048608,-2147450848,0,-2147483648,32768,1081376,-2146435072,1048608,-2147483616,0,1081344,32800,-2146402304,-2146435072,32800,0,1081376,-2146435040,1048576,-2147450848,-2146435072,-2146402304,32768,-2146435072,-2147450880,32,-2146402272,1081376,32,32768,-2147483648,32800,-2146402304,1048576,-2147483616,1048608,-2147450848,-2147483616,1048608,1081344,0,-2147450880,32800,-2147483648,-2146435040,-2146402272,1081344);
        var spfunction3 = new Array(520,134349312,0,134348808,134218240,0,131592,134218240,131080,134217736,134217736,131072,134349320,131080,134348800,520,134217728,8,134349312,512,131584,134348800,134348808,131592,134218248,131584,131072,134218248,8,134349320,512,134217728,134349312,134217728,131080,520,131072,134349312,134218240,0,512,131080,134349320,134218240,134217736,512,0,134348808,134218248,131072,134217728,134349320,8,131592,131584,134217736,134348800,134218248,520,134348800,131592,8,134348808,131584);
        var spfunction4 = new Array(8396801,8321,8321,128,8396928,8388737,8388609,8193,0,8396800,8396800,8396929,129,0,8388736,8388609,1,8192,8388608,8396801,128,8388608,8193,8320,8388737,1,8320,8388736,8192,8396928,8396929,129,8388736,8388609,8396800,8396929,129,0,0,8396800,8320,8388736,8388737,1,8396801,8321,8321,128,8396929,129,1,8192,8388609,8193,8396928,8388737,8193,8320,8388608,8396801,128,8388608,8192,8396928);
        var spfunction5 = new Array(256,34078976,34078720,1107296512,524288,256,1073741824,34078720,1074266368,524288,33554688,1074266368,1107296512,1107820544,524544,1073741824,33554432,1074266112,1074266112,0,1073742080,1107820800,1107820800,33554688,1107820544,1073742080,0,1107296256,34078976,33554432,1107296256,524544,524288,1107296512,256,33554432,1073741824,34078720,1107296512,1074266368,33554688,1073741824,1107820544,34078976,1074266368,256,33554432,1107820544,1107820800,524544,1107296256,1107820800,34078720,0,1074266112,1107296256,524544,33554688,1073742080,524288,0,1074266112,34078976,1073742080);
        var spfunction6 = new Array(536870928,541065216,16384,541081616,541065216,16,541081616,4194304,536887296,4210704,4194304,536870928,4194320,536887296,536870912,16400,0,4194320,536887312,16384,4210688,536887312,16,541065232,541065232,0,4210704,541081600,16400,4210688,541081600,536870912,536887296,16,541065232,4210688,541081616,4194304,16400,536870928,4194304,536887296,536870912,16400,536870928,541081616,4210688,541065216,4210704,541081600,0,541065232,16,16384,541065216,4210704,16384,4194320,536887312,0,541081600,536870912,4194320,536887312);
        var spfunction7 = new Array(2097152,69206018,67110914,0,2048,67110914,2099202,69208064,69208066,2097152,0,67108866,2,67108864,69206018,2050,67110912,2099202,2097154,67110912,67108866,69206016,69208064,2097154,69206016,2048,2050,69208066,2099200,2,67108864,2099200,67108864,2099200,2097152,67110914,67110914,69206018,69206018,2,2097154,67108864,67110912,2097152,69208064,2050,2099202,69208064,2050,67108866,69208066,69206016,2099200,0,2,69208066,0,2099202,69206016,2048,67108866,67110912,2048,2097154);
        var spfunction8 = new Array(268439616,4096,262144,268701760,268435456,268439616,64,268435456,262208,268697600,268701760,266240,268701696,266304,4096,64,268697600,268435520,268439552,4160,266240,262208,268697664,268701696,4160,0,0,268697664,268435520,268439552,266304,262144,266304,262144,268701696,4096,64,268697664,4096,266304,268439552,64,268435520,268697600,268697664,268435456,262144,268439616,0,268701760,262208,268435520,268697600,268439552,268439616,0,268701760,266240,266240,4160,4160,262208,268435456,268701696);
        var keys = $.su.des_createKeys(key);
        var m = 0, i, j, temp, temp2, right1, right2, left, right, looping;
        var cbcleft, cbcleft2, cbcright, cbcright2;
        var endloop, loopinc;
        var len = message.length;
        var chunk = 0;
        var iterations = keys.length == 32 ? 3 : 9;
        if (iterations == 3) {
            looping = encrypt ? new Array(0,32,2) : new Array(30,-2,-2)
        } else {
            looping = encrypt ? new Array(0,32,2,62,30,-2,64,96,2) : new Array(94,62,-2,32,64,2,30,-2,-2)
        }
        if (padding == 2) {
            message += "        "
        } else {
            if (padding == 1) {
                if (encrypt) {
                    temp = 8 - (len % 8);
                    message += String.fromCharCode(temp, temp, temp, temp, temp, temp, temp, temp);
                    if (temp === 8) {
                        len += 8
                    }
                }
            } else {
                if (!padding) {
                    message += "\0\0\0\0\0\0\0\0"
                }
            }
        }
        var result = "";
        var tempresult = "";
        if (mode == 1) {
            cbcleft = (iv.charCodeAt(m++) << 24) | (iv.charCodeAt(m++) << 16) | (iv.charCodeAt(m++) << 8) | iv.charCodeAt(m++);
            cbcright = (iv.charCodeAt(m++) << 24) | (iv.charCodeAt(m++) << 16) | (iv.charCodeAt(m++) << 8) | iv.charCodeAt(m++);
            m = 0
        }
        while (m < len) {
            left = (message.charCodeAt(m++) << 24) | (message.charCodeAt(m++) << 16) | (message.charCodeAt(m++) << 8) | message.charCodeAt(m++);
            right = (message.charCodeAt(m++) << 24) | (message.charCodeAt(m++) << 16) | (message.charCodeAt(m++) << 8) | message.charCodeAt(m++);
            if (mode == 1) {
                if (encrypt) {
                    left ^= cbcleft;
                    right ^= cbcright
                } else {
                    cbcleft2 = cbcleft;
                    cbcright2 = cbcright;
                    cbcleft = left;
                    cbcright = right
                }
            }
            temp = ((left >>> 4) ^ right) & 252645135;
            right ^= temp;
            left ^= (temp << 4);
            temp = ((left >>> 16) ^ right) & 65535;
            right ^= temp;
            left ^= (temp << 16);
            temp = ((right >>> 2) ^ left) & 858993459;
            left ^= temp;
            right ^= (temp << 2);
            temp = ((right >>> 8) ^ left) & 16711935;
            left ^= temp;
            right ^= (temp << 8);
            temp = ((left >>> 1) ^ right) & 1431655765;
            right ^= temp;
            left ^= (temp << 1);
            left = ((left << 1) | (left >>> 31));
            right = ((right << 1) | (right >>> 31));
            for (j = 0; j < iterations; j += 3) {
                endloop = looping[j + 1];
                loopinc = looping[j + 2];
                for (i = looping[j]; i != endloop; i += loopinc) {
                    right1 = right ^ keys[i];
                    right2 = ((right >>> 4) | (right << 28)) ^ keys[i + 1];
                    temp = left;
                    left = right;
                    right = temp ^ (spfunction2[(right1 >>> 24) & 63] | spfunction4[(right1 >>> 16) & 63] | spfunction6[(right1 >>> 8) & 63] | spfunction8[right1 & 63] | spfunction1[(right2 >>> 24) & 63] | spfunction3[(right2 >>> 16) & 63] | spfunction5[(right2 >>> 8) & 63] | spfunction7[right2 & 63])
                }
                temp = left;
                left = right;
                right = temp
            }
            left = ((left >>> 1) | (left << 31));
            right = ((right >>> 1) | (right << 31));
            temp = ((left >>> 1) ^ right) & 1431655765;
            right ^= temp;
            left ^= (temp << 1);
            temp = ((right >>> 8) ^ left) & 16711935;
            left ^= temp;
            right ^= (temp << 8);
            temp = ((right >>> 2) ^ left) & 858993459;
            left ^= temp;
            right ^= (temp << 2);
            temp = ((left >>> 16) ^ right) & 65535;
            right ^= temp;
            left ^= (temp << 16);
            temp = ((left >>> 4) ^ right) & 252645135;
            right ^= temp;
            left ^= (temp << 4);
            if (mode == 1) {
                if (encrypt) {
                    cbcleft = left;
                    cbcright = right
                } else {
                    left ^= cbcleft2;
                    right ^= cbcright2
                }
            }
            tempresult += String.fromCharCode((left >>> 24), ((left >>> 16) & 255), ((left >>> 8) & 255), (left & 255), (right >>> 24), ((right >>> 16) & 255), ((right >>> 8) & 255), (right & 255));
            chunk += 8;
            if (chunk == 512) {
                result += tempresult;
                tempresult = "";
                chunk = 0
            }
        }
        result += tempresult;
        result = result.replace(/\0*$/g, "");
        if (!encrypt) {
            if (padding === 1) {
                var len = result.length
                  , paddingChars = 0;
                len && (paddingChars = result.charCodeAt(len - 1));
                (paddingChars <= 8) && (result = result.substring(0, len - paddingChars))
            }
            result = decodeURIComponent(escape(result))
        }
        return result
    }
    ;
    $.su.des_createKeys = function(key) {
        var pc2bytes0 = new Array(0,4,536870912,536870916,65536,65540,536936448,536936452,512,516,536871424,536871428,66048,66052,536936960,536936964);
        var pc2bytes1 = new Array(0,1,1048576,1048577,67108864,67108865,68157440,68157441,256,257,1048832,1048833,67109120,67109121,68157696,68157697);
        var pc2bytes2 = new Array(0,8,2048,2056,16777216,16777224,16779264,16779272,0,8,2048,2056,16777216,16777224,16779264,16779272);
        var pc2bytes3 = new Array(0,2097152,134217728,136314880,8192,2105344,134225920,136323072,131072,2228224,134348800,136445952,139264,2236416,134356992,136454144);
        var pc2bytes4 = new Array(0,262144,16,262160,0,262144,16,262160,4096,266240,4112,266256,4096,266240,4112,266256);
        var pc2bytes5 = new Array(0,1024,32,1056,0,1024,32,1056,33554432,33555456,33554464,33555488,33554432,33555456,33554464,33555488);
        var pc2bytes6 = new Array(0,268435456,524288,268959744,2,268435458,524290,268959746,0,268435456,524288,268959744,2,268435458,524290,268959746);
        var pc2bytes7 = new Array(0,65536,2048,67584,536870912,536936448,536872960,536938496,131072,196608,133120,198656,537001984,537067520,537004032,537069568);
        var pc2bytes8 = new Array(0,262144,0,262144,2,262146,2,262146,33554432,33816576,33554432,33816576,33554434,33816578,33554434,33816578);
        var pc2bytes9 = new Array(0,268435456,8,268435464,0,268435456,8,268435464,1024,268436480,1032,268436488,1024,268436480,1032,268436488);
        var pc2bytes10 = new Array(0,32,0,32,1048576,1048608,1048576,1048608,8192,8224,8192,8224,1056768,1056800,1056768,1056800);
        var pc2bytes11 = new Array(0,16777216,512,16777728,2097152,18874368,2097664,18874880,67108864,83886080,67109376,83886592,69206016,85983232,69206528,85983744);
        var pc2bytes12 = new Array(0,4096,134217728,134221824,524288,528384,134742016,134746112,16,4112,134217744,134221840,524304,528400,134742032,134746128);
        var pc2bytes13 = new Array(0,4,256,260,0,4,256,260,1,5,257,261,1,5,257,261);
        var iterations = key.length > 8 ? 3 : 1;
        var keys = new Array(32 * iterations);
        var shifts = new Array(0,0,1,1,1,1,1,1,0,1,1,1,1,1,1,0);
        var lefttemp, righttemp, m = 0, n = 0, temp;
        for (var j = 0; j < iterations; j++) {
            var left = (key.charCodeAt(m++) << 24) | (key.charCodeAt(m++) << 16) | (key.charCodeAt(m++) << 8) | key.charCodeAt(m++);
            var right = (key.charCodeAt(m++) << 24) | (key.charCodeAt(m++) << 16) | (key.charCodeAt(m++) << 8) | key.charCodeAt(m++);
            temp = ((left >>> 4) ^ right) & 252645135;
            right ^= temp;
            left ^= (temp << 4);
            temp = ((right >>> -16) ^ left) & 65535;
            left ^= temp;
            right ^= (temp << -16);
            temp = ((left >>> 2) ^ right) & 858993459;
            right ^= temp;
            left ^= (temp << 2);
            temp = ((right >>> -16) ^ left) & 65535;
            left ^= temp;
            right ^= (temp << -16);
            temp = ((left >>> 1) ^ right) & 1431655765;
            right ^= temp;
            left ^= (temp << 1);
            temp = ((right >>> 8) ^ left) & 16711935;
            left ^= temp;
            right ^= (temp << 8);
            temp = ((left >>> 1) ^ right) & 1431655765;
            right ^= temp;
            left ^= (temp << 1);
            temp = (left << 8) | ((right >>> 20) & 240);
            left = (right << 24) | ((right << 8) & 16711680) | ((right >>> 8) & 65280) | ((right >>> 24) & 240);
            right = temp;
            for (var i = 0; i < shifts.length; i++) {
                if (shifts[i]) {
                    left = (left << 2) | (left >>> 26);
                    right = (right << 2) | (right >>> 26)
                } else {
                    left = (left << 1) | (left >>> 27);
                    right = (right << 1) | (right >>> 27)
                }
                left &= -15;
                right &= -15;
                lefttemp = pc2bytes0[left >>> 28] | pc2bytes1[(left >>> 24) & 15] | pc2bytes2[(left >>> 20) & 15] | pc2bytes3[(left >>> 16) & 15] | pc2bytes4[(left >>> 12) & 15] | pc2bytes5[(left >>> 8) & 15] | pc2bytes6[(left >>> 4) & 15];
                righttemp = pc2bytes7[right >>> 28] | pc2bytes8[(right >>> 24) & 15] | pc2bytes9[(right >>> 20) & 15] | pc2bytes10[(right >>> 16) & 15] | pc2bytes11[(right >>> 12) & 15] | pc2bytes12[(right >>> 8) & 15] | pc2bytes13[(right >>> 4) & 15];
                temp = ((righttemp >>> 16) ^ lefttemp) & 65535;
                keys[n++] = lefttemp ^ temp;
                keys[n++] = righttemp ^ (temp << 16)
            }
        }
        return keys
    }
    ;
    $.su.genkey = function(key, start, end) {
        return {
            key: $.su.pad(key.slice(start, end)),
            vector: 1
        }
    }
    ;
    $.su.pad = function(key) {
        for (var i = key.length; i < 24; i++) {
            key += "0"
        }
        return key
    }
    ;
    $.su.DES3 = {
        encrypt: function(input) {
            var genKey = $.su.genkey("PKCS5Padding", 0, 24);
            return btoa($.su.des(genKey.key, input, 1, 1, "26951234", 1))
        },
        decrypt: function(input) {
            var genKey = $.su.genkey("PKCS5Padding", 0, 24);
            return $.su.des(genKey.key, atob(input), 0, 1, "26951234", 1)
        }
    }
}
)(jQuery);

```

AlexandrErohin commented:
Thank you very much! I have one comment for encrypted password only

AlexandrErohin commented:
Does this steps works for you to get the hashedPassword ?

1. Go to the login page of your router. (default: 192.168.0.1)
2. Type in the password you use to login into the password field.
3. Click somewhere else on the page so that the password field is not selected anymore.
4. Open the JavaScript console of your browser (usually by pressing F12 and then clicking on "Console").
5. Type `document.getElementById("login-password").value;` or `document.getElementById("pcPassword").value;`, depending on your firmware version.
6. Copy the returned value as password.

swwgames commented:
yes that worked, document.getElementById("login-password").value;  gave me the value.

AlexandrErohin commented:
@swwgames Hi. Could you test please v3.4.0?

swwgames commented:
no that version doesn't work.
what i needed to change for it to work was:
```
class TplinkC1200Router(TplinkBaseRouter):
    def supports(self) -> bool:
        return True

    def authorize(self) -> None:
        if len(self.password) < 200:
            raise Exception('You need to use web encrypted password instead. Check the documentation!')

        url = '{}/cgi-bin/luci/;stok=/login?form=login'.format(self.host)

        response = requests.post(
            url,
            params={'operation': 'login', 'username': self.username, 'password': self.password},
            #headers={'Referer': self._login_referer, 'Content-Type': 'application/x-www-form-urlencoded'},
            timeout=self.timeout,
            verify=self._verify_ssl,
        )
```
Here i removed the headers. the _login_referer is empty so it did not work.

And Here i removed the &operation=read
```
class TplinkBaseRouter(AbstractRouter, TplinkRequest):
    def __init__(self, host: str, password: str, username: str = 'admin', logger: Logger = None,
                 verify_ssl: bool = True, timeout: int = 10) -> None:
        super().__init__(host, password, username, logger, verify_ssl, timeout)

        self._url_firmware = 'admin/firmware?form=upgrade&operation=read'
        self._url_wireless_stats = 'admin/wireless?form=statistics'
        self._url_ipv4_reservations = 'admin/dhcps?form=reservation&operation=load'
        self._url_ipv4_dhcp_leases = 'admin/dhcps?form=client&operation=load'
```
i also needed to change in the baserouter class in the get_status the operation from read to load.
```
        for item in self.request(self._url_wireless_stats, 'operation=load'):
            if item['mac'] not in devices:
                status.wifi_clients_total += 1
                type = self._map_wire_type(item.get('type'))
                devices[item['mac']] = Device(type, macaddress.EUI48(item['mac']), ipaddress.IPv4Address('0.0.0.0'), '')
            devices[item['mac']].packets_sent = item.get('txpkts')
            devices[item['mac']].packets_received = item.get('rxpkts')
```
first it was sending twice operation read, now it only sends operation load. and i get the data.

feel free to ask further questions.

AlexandrErohin commented:
@swwgames Thank you!
I have fixed it. Could you try v3.4.3 and let me know please?

swwgames commented:
Yes it work!

AlexandrErohin commented:
@swwgames Thank you!
